### ISSUE TITLE
Tips on how to debug cache

### ISSUE THREAD

**Issue by Sewdn** [zino-app/graphql-flutter#383]
Hi!

We are currently struggling with implementing offline experience reading all queries with cacheOnly fetchPolicy and only using cacheAndNetwork policy if it is certain we have connection (and service is available).

Could someone give advice on
- how to inspect the current runTime cache?
- how to inspect the device's persistent offline storage cache?

Furthermore, we sometimes notice QueryResultSource Loading being used when we expect Cache as the result source. We suspect this is only during debug. What is the Loading QueryResultSource? We would expect QueryResultSource being Cache or Network (and result.loading being true or false when Network and false when cache).

**Comment by micimize** [comment]
> how to inspect the current runTime cache?

My approach is to `import` whatever file my cache is defined in, and then set a breakpoint where I want to do comparisons or have cache keys handy. Unfortunately the [`LazyCacheMap`](https://github.com/zino-app/graphql-flutter/blob/master/packages/graphql/lib/src/cache/lazy_cache_map.dart) is a bit hard to navigate due to lack of a good `toString`, but you can always look at it's raw `.data`   

> how to inspect the device's persistent offline storage cache?

Not sure, would be handy though

<br/>

> What is the Loading QueryResultSource?

`QueryResultSource.Loading` was created when cache hits still had a `loading` message sent beforehand. Maybe it makes less sense now, I'm not sure. 

Debugging shouldn't matter for the `loading` state. You should only see the loading state in your situation when using `cacheAndNetwork` and there's a cache miss. A cache miss with `cacheOnly` should fail with an error.

**Comment by micimize** [comment]
Here are some helpers for printing out `LazyCacheMaps`:
```dart
import 'package:flutter/material.dart' show debugPrint;
import 'package:graphql/src/utilities/traverse.dart';
import 'dart:convert';

import 'package:graphql_flutter/graphql_flutter.dart' show LazyCacheMap;

final JsonEncoder encoder = const JsonEncoder.withIndent('  ');

class LazyCacheMapTraversal extends Traversal {
  LazyCacheMapTraversal() : super((Object n) => null);

  @override
  Object transform(Object node) {
    if (node is LazyCacheMap) {
      if (alreadySeen(node)) {
        return '...';
      }
      final map = Map<String, dynamic>.fromEntries(node.entries);

      return traverseValues(map);
    }
    return null;
  }
}

dynamic getJson(LazyCacheMap cacheMap) {
  return LazyCacheMapTraversal().traverse(cacheMap);
}

void pprint(dynamic json) {
  if (json is LazyCacheMap) {
    json = getJson(json);
  }
  debugPrint(encoder.convert(json));
}
```

**Comment by micimize** [comment]
I wrote some helpers for printing `LazyCacheMaps`, and in the process discovered some bugs. So, the following will only work safely with #399:
```dart
import 'package:flutter/material.dart' show debugPrint;
import 'package:graphql/src/utilities/traverse.dart';
import 'dart:convert';

import 'package:graphql_flutter/graphql_flutter.dart' show LazyCacheMap;

final JsonEncoder encoder = const JsonEncoder.withIndent('  ');

class LazyCacheMapTraversal extends Traversal {
  LazyCacheMapTraversal() : super((Object n) => null);

  Object traverse(Object node) {
    final seen = alreadySeen(node);
    if (node is LazyCacheMap) {
      if (seen) {
        return '...';
      }
      node = Map<String, dynamic>.fromEntries((node as LazyCacheMap).entries);
    }
    if (seen) {
      return node;
    }
    if (node is List<Object>) {
      return node.map<Object>((Object node) => traverse(node)).toList();
    }
    if (node is Map<String, Object>) {
      return traverseValues(node);
    }
    return node;
  }
}

dynamic getJson(LazyCacheMap cacheMap) {
  return LazyCacheMapTraversal().traverse(cacheMap);
}

void pprint(dynamic json) {
  if (json is LazyCacheMap) {
    json = getJson(json);
  }
  debugPrint(encoder.convert(json));
}
```

**Comment by micimize** [comment]
There was a loading bug that was recently fixed, which may have caused confusion (#351), and #398 might also be helpful for your use case.
